Parser: Fix location (line, col) for diagnostics

nanoant opened the pull request:
Previously parser was using lexMessage which was taking location from current
buffer position which was pointing after recently consumed token. But since
parser shows diagnostics about that token it should point to the location where
token starts.

This makes diagnostics like: `test.nim(2, 2) Error: ':' expected` point
properly at the beginning of the wrong token.

Araq commented on compiler/parser.nim:
```diff
@@ -91,7 +91,7 @@ proc closeParser(p: var TParser) =
 proc parMessage(p: TParser, msg: TMsgKind, arg = "") =
   ## Produce and emit the parser message `arg` to output.
-  lexMessage(p.lex, msg, arg)
+  lexMessageTok(p.lex, msg, p.tok, arg)
```
Is this correct in any case? Also for errInvalidIndentation?

nanoant replied:
Very good question. I think this is correct. I've looked into a parser and `errInvalidIndentation` is caused by wrong `p.tok.indent`, so the this error applies to the token itself, so we expect compiler point on the token, not `l.bufpos` which points on the 1st char after the token.
Actually parser always work on the token not on the characters in general, so it is okay that lexer use `l.bufpos` to determine problematic place, but for parser it is IMHO always location of token.

nanoant replied:
Also I think (if there are not already) maybe it would be worth to have some tests on expected error messages. Just to be sure that similar commit don't cause a regression. I think I will ping you on IRC for some questions about tests, especially these failing compilation.

nanoant commented:
@Araq 

1. Rebased to `devel`
2. Fixed incorrect column numbering, should start from 1 - same as lines
3. Added extra tests for some compiler diagnostics (this one depends on PR https://github.com/Araq/Nim/pull/2575)

Araq commented:
Columns *need* to start from 0. This is documented and lots of tools rely on this.

Araq commented on compiler/msgs.nim:
```diff
@@ -804,8 +804,11 @@ proc liMessage(info: TLineInfo, msg: TMsgKind, arg: string,
+  # first column to be 1, so we need to +1 here
   let s = frmt % [toMsgFilename(info), coordToStr(info.line),
-                  coordToStr(info.col), getMessageStr(msg, arg)]
+                  coordToStr(info.col+1), getMessageStr(msg, arg)]
```
That's not the only place where coordToStr is called with a column.

nanoant replied:
I know, but it is only place where actual message going out from compiler is generated. Intentionally ignored `proc formatMsg*` which is used only in `vm` internals. Should or shouldn't I change `formatMsg` too?

Araq commented:
there is also writeContext which needs to be updated. And yes, update formatMsg too please.
